### PR TITLE
Split caikit-tgis into separate containers

### PR DESCRIPTION
**Pull request by dtrifiro**
## Description
- add `Dockerfile` for `caikit-nlp` with `tgis` backend
- update docs example (`InferenceService`+`ServingRuntime`)
- remove smoketest (this breaks with the new caikit-only image

## How Has This Been Tested?

This can be tested by running the provided example in either a local kserve deployment or openshift:

This example uses a pvc/pvc, but the model can be also stored in s3 or other object storage

1. Create a PV/PVC
```yaml
apiVersion: v1
kind: PersistentVolume
metadata:
  name: caikit-volume
  labels:
    type: local
spec:
  storageClassName: manual
  capacity:
    storage: 5Gi
  accessModes:
    - ReadWriteMany
  hostPath:
    path: "/mnt/models/"
---
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: caikit-claim
spec:
  storageClassName: manual
  accessModes:
    - ReadWriteMany
  resources:
    requests:
      storage: 2Gi
```

Download a model and convert it to `caikit` format (note: this requires `git-lfs` to be installed and configured
```bash
git clone https://huggingface.co/google/flan-t5-small
python -c 'import caikit_nlp; \
model = caikit_nlp.text_generation.TextGeneration.bootstrap("flan-t5-small"); \
model.save("flan-t5-small-caikit")'
```

2. Copy the model to the pv

3. Edit `docs/caikit-isvc.yaml` and make `storageUri` point to the model:
`storageUri: pvc://caikit-claim/flan-t5-small-caikit/`

3 Apply the manifests,
```bash
oc apply -f docs/caikit-isvc.yaml -f docs/caikit-servingruntime.yaml
```
4. After port forwarding to the inference service, it should be able to query the grpc endpoint
```bash
oc port-forward deployments/caikit-example-isvc-predictor-00001-deployment 8085 &>/dev/null &
grpcurl -plaintext -d '{"text": "At what temperature does Nitrogen boil?"}' -H "mm-model-id: flan-t5-small-caikit" 127.0.0.1:8085 caikit.runtime.Nlp.NlpService/TextGenerationTaskPredict
```
Verify that the inference is running on the tgis instance by looking at the logs of the transformer container:
```bash
oc logs deployments/caikit-example-isvc-predictor-00001-deployment -c transformer-container -f     
```

In a subsequent PR, I'll provide an automatic method of testing this setup, either via docker-compose and/or by deploying these manifests.

## Merge criteria:
- [x] The commits are squashed in a cohesive manner and have meaningful messages.
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

File: Dockerfile
Comment (spolti) [inline Dockerfile:1]: I think in this case is safe to use latest tag.

File: Dockerfile
Comment (spolti) [inline Dockerfile:2]: can't we use the ubi8-minimal or micro?

File: Dockerfile
Comment (spolti) [inline Dockerfile:36]: 2 cents for improve the user addition:

```
groupadd -r caikit -g 1001 && useradd -u 1001 -r -g root -G "caikit" -m -d "/caikit" -s /sbin/nologin -c "Caikit User  user" caikit
```

ps. this runs well with ocp random uid as well.

Comment (dtrifiro) [reply Dockerfile:36]: Any particular reason for having `-g root`?

Comment (spolti) [reply Dockerfile:36]: better readability, it is the same than `-g 0`

Comment (Xaenalt) [reply Dockerfile:36]: There's some openshift-specific semantics around this, I think the symbolic name works, but the group 0 assigned I don't think is quite the same as the group that root is a part of, there's a doc about this somewhere with the exact behavior on openshift

Comment (dtrifiro) [reply Dockerfile:36]: > better readability, it is the same than -g 0

Sorry, I meant if there's any specific reason for adding the `caikit` user to the root group, maybe it's just better to get rid of it now that I see it.

Comment (kpouget) [reply Dockerfile:36]: @dtrifiro, it's the solution to support OpenShift arbitrary user ID, see there:
> https://docs.openshift.com/container-platform/4.11/openshift_images/create-images.html#use-uid_create-images
```
RUN chgrp -R 0 /some/directory && \
    chmod -R g=u /some/directory
```
that's the way to give read/write access to `anyuid` user, contrary to the more `chown 1001`  which implies that the container runs with user `1001`

File: Makefile
Comment (z103cb) [inline Makefile:4]: Does it make sense to allow th users to use something other than podman ?

```suggestion
# Container Engine to be used for building images
ENGINE ?= podman

default:
	$(ENGINE) build \
```

File: Makefile
Comment (z103cb) [inline Makefile:10]: This target will fail if `opendatahub/caikit:latest` is not in the user's local repo.
```suggestion
# This target requires that the opendatahub/caikit:latest is available in your local registry.
#  you may execute podman pull opendatahub/caikit:latest or make default. 
refresh-poetry-lock-files:
```

File: demo/kserve/built-tip.md
Comment (z103cb) [inline demo/kserve/built-tip.md:8]: Is there a macos equivalent ? Should we mention if you are a "MacOS" user you are SOL ?

File: Makefile
Comment (z103cb) [inline Makefile:5]: On an M1 Mac and podman this fails with this error:

```text
copying psutil/tests/test_windows.py -> build/lib.linux-aarch64-cpython-39/psutil/tests
  running build_ext
  building 'psutil._psutil_linux' extension
  creating build/temp.linux-aarch64-cpython-39
  creating build/temp.linux-aarch64-cpython-39/psutil
  gcc -pthread -Wno-unused-result -Wsign-compare -DDYNAMIC_ANNOTATIONS_ENABLED=1 -DNDEBUG -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -fasynchronous-unwind-tables -fstack-clash-protection -D_GNU_SOURCE -fPIC -fwrapv -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -fasynchronous-unwind-tables -fstack-clash-protection -D_GNU_SOURCE -fPIC -fwrapv -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -fasynchronous-unwind-tables -fstack-clash-protection -D_GNU_SOURCE -fPIC -fwrapv -fPIC -DPSUTIL_POSIX=1 -DPSUTIL_SIZEOF_PID_T=4 -DPSUTIL_VERSION=595 -DPy_LIMITED_API=0x03060000 -DPSUTIL_ETHTOOL_MISSING_TYPES=1 -DPSUTIL_LINUX=1 -I/tmp/tmpcp388sgs/.venv/include -I/usr/include/python3.9 -c psutil/_psutil_common.c -o build/temp.linux-aarch64-cpython-39/psutil/_psutil_common.o
  psutil could not be installed from sources because gcc is not installed. Try running:
    sudo yum install gcc python3-devel
  error: command 'gcc' failed: No such file or directory
  

  at /usr/local/lib/python3.9/site-packages/poetry/installation/chef.py:147 in _prepare
      143│ 
      144│                 error = ChefBuildError("\n\n".join(message_parts))
      145│ 
      146│             if error is not None:
    → 147│                 raise error from None
      148│ 
      149│             return path
      150│ 
      151│     def _prepare_sdist(self, archive: Path, destination: Path | None = None) -> Path:

Note: This error originates from the build backend, and is likely not a problem with poetry but with psutil (5.9.5) not supporting PEP 517 builds. You can verify this by running 'pip wheel --use-pep517 "psutil (==5.9.5)"'.

Error: building at STEP "RUN pip3 install poetry && poetry install": while running runtime: exit status 1

```

File: Makefile
Comment (dtrifiro) [reply Makefile:10]: Forgot to add `: default` dependency. Nice catch

File: demo/kserve/built-tip.md
Comment (dtrifiro) [reply demo/kserve/built-tip.md:8]: Well, I'm assuming the equivalent is `brew install git-lfs`

File: Dockerfile
Comment (z103cb) [inline Dockerfile:14]: This seems to have gotten me past the issue have reported on the makefile
```suggestion
        git git-lfs python39-pip python39-wheel gcc python3-devel && \
```

Comment (dtrifiro) [reply Dockerfile:14]: The fact that `gcc` is required means that the wheels are not available and have to be built (using gcc). I think this is only required for arm (aarch64) targets, I'd rather avoid adding `gcc` to the base image.

File: Dockerfile
Comment (z103cb) [inline Dockerfile:9]: Would it be possible to determine the CPU arch at build time and use the appropiate package? On arm64 this seems to be the right package.
```suggestion
RUN curl -Lo /usr/local/bin/grpc_health_probe https://github.com/grpc-ecosystem/grpc-health-probe/releases/download/${GRPC_HEALTH_PROBE_VERSION}/grpc_health_probe-linux-arm64 && \
```

File: Dockerfile
Comment (dtrifiro) [reply Dockerfile]: Updated using long options.

File: Dockerfile
Comment (dtrifiro) [reply Dockerfile]: Yep, moved to using `minimal`

File: Makefile
Comment (dtrifiro) [reply Makefile:4]: Wondering the same, but for now I'd leave this as is, possibly adding this feature in another PR.

File: Dockerfile
Comment (spolti) [reply Dockerfile]: @dtrifiro sry, I miss understand your question, I thought as you were adding the user to the root group already you knew it was necessary as @kpouget mentioned.

File: Dockerfile
Comment (Xaenalt) [inline Dockerfile:45]: Let's also see about moving to the combined HTTP+gRPC entrypoint Gabe was talking about here, there should be a caikit.runtime entrypoint that launches both in recent caikit (may need to bump version to see it)

File: Dockerfile
Comment (Xaenalt) [reply Dockerfile:13]: I tend to agree here, we want the image to be very minimal. If there are python packages that we don't participate in and can't push arm wheels up to pypi, we may need a builder image for when we go to multiarch, but the release image shouldn't have gcc and dev tools per standard

File: demo/kserve/built-tip.md
Comment (Xaenalt) [reply demo/kserve/built-tip.md:8]: https://docs.github.com/en/repositories/working-with-files/managing-large-files/installing-git-large-file-storage?platform=mac

File: pyproject.toml
Comment (Xaenalt) [inline pyproject.toml:13]: Love this comment

File: docs/caikit-servingruntime.yaml
Comment (Xaenalt) [inline docs/caikit-servingruntime.yaml:40]: This is currently still the case for the combined container, as well as caikit standalone. TGIS standalone can serve more standard model formats

File: docs/caikit-servingruntime.yaml
Comment (Xaenalt) [inline docs/caikit-servingruntime.yaml:45]: These two are named backwards by paradigm I think, typically a transformer in the context of standard model serving mutates the input, which is what caikit will be doing

Additionally, caikit doesn't need the TRANSFORMERS_CACHE variable, only TGIS. We may need to revisit that path later as well because that's where it searches for any pre-downloaded HF models. For now though, not needed for the caikit side

File: docs/caikit-servingruntime.yaml
Comment (Xaenalt) [inline docs/caikit-servingruntime.yaml:7]: This is a great idea, love it

File: Dockerfile
Comment (Xaenalt) [inline Dockerfile:27]: Oh, actually upon reading deeper here, yeah, we could sidestep holding some of the holding onto stuff like devel pkgs by changing this to be the minimal container, rather than chaining on top of base

This is also good for cleaning out things like the poetry lock from the shipping image

File: Makefile
Comment (Xaenalt) [reply Makefile:4]: Yep, future PR

File: Makefile
Comment (Xaenalt) [reply Makefile:5]: Yeah, building in the image for multiarch will have to come separately

File: docs/caikit-servingruntime.yaml
Comment (Xaenalt) [inline docs/caikit-servingruntime.yaml:60]: TRANSFORMERS_CACHE only needed for TGIS, not caikit

File: docs/caikit-servingruntime.yaml
Comment (Xaenalt) [inline docs/caikit-servingruntime.yaml:51]: For TGIS, I want to add the following into the template to align with what perf/scale has been testing, these are probably fine defaults:

```
    - name: RUNTIME_GRPC_SERVER_THREAD_POOL_SIZE
      value: "64" # This may need to scale higher, some models have this at 256, but this seems like a good default up from 5
    - name: MAX_BATCH_SIZE
      value: "256"
    - name: MAX_CONCURRENT_REQUESTS
      value: "64"
    - name: MAX_BATCH_WEIGHT
      value: "10000"
```

And a couple others that we may want to comment out, but leave in the SR for doc purposes

```
    - name: PT2_COMPILE
      value: "true" # This makes the model take a lot of time to load, but provides a speedup 
    - name: FLASH_ATTENTION
      value: "true" # This optimizes a few models that need it, should be commented by default with a link to https://github.com/IBM/text-generation-inference#converting-weights-to-safetensors-format
```

For sharding, we also need these fields, and *might* need to update the load_timeout in caikit, but I think that's likely not a problem in remote mode

Ref: https://github.com/IBM/text-generation-inference#running-sharded-models-tensor-parallel

```
    - name: NUM_GPUS
      value: "4" # Presumably however many they have
    - name: DEPLOYMENT_FRAMEWORK
      value: hf_custom_tp
```

It also pairs with the additional last line in the following caikit runtime-config.yaml:

```
    model_management:
      initializers:
        default:
          type: LOCAL
          config:
            backend_priority:
              - type: TGIS
                config:
                  local:
                    load_timeout: 2000
                    num_gpus: 4
```

Comment (Xaenalt) [reply docs/caikit-servingruntime.yaml:51]: Given the potential size of the changes for sharding, it may be worth deferring that to a second SR spec that's set up specifically for it

Comment (Xaenalt) [reply docs/caikit-servingruntime.yaml:51]: Also investigating if NUM_GPUs and similar need to be set explicitly or if there's detection logic

File: Dockerfile
Comment (dtrifiro) [reply Dockerfile:45]: How about doing this in another PR?

File: Dockerfile
Comment (dtrifiro) [reply Dockerfile:8]: @Xaenalt  how about getting rid of this?

File: docs/caikit-servingruntime.yaml
Comment (dtrifiro) [reply docs/caikit-servingruntime.yaml:51]: Regarding sharding: I'd rather create a new manifest for it and include in another PR

File: docs/caikit-servingruntime.yaml
Comment (dtrifiro) [reply docs/caikit-servingruntime.yaml]: yep, removed.

File: docs/caikit-servingruntime.yaml
Comment (dtrifiro) [reply docs/caikit-servingruntime.yaml:59]: Regarding `max_batch_weight=10000`: it seems it is not a safe default, resulting in a tgis panic:

> thread 'main' panicked at 'max_batch_weight (10000) not large enough for max_sequence_length (2048)', /usr/src/router/src/server.rs:217:17

I'd comment this out for now, see #112 to reproduce

File: Dockerfile
Comment (Xaenalt) [reply Dockerfile:8]: Yeah, we should take that down as a note for when we go to multiarch builds, but yeah separate PR for now is fine

File: Dockerfile
Comment (Xaenalt) [reply Dockerfile:27]: So when we do from base as deploy, I may be wrong, but I believe it keeps everything from the previous layers that were created, so we'd have things like `poetry.lock` in the end container. In general, best practice is probably not to have that type of file in the final build, and we're already copying --from poetry-builder, so we could instead do ubi-minimal as deploy, and whatever you want as base since we copy the .venv anyway

File: Dockerfile
Comment (Xaenalt) [reply Dockerfile:45]: Sure, that works, we'll need that one as next priority though so we have time to play with the HTTP API

File: docs/caikit-servingruntime.yaml
Comment (Xaenalt) [reply docs/caikit-servingruntime.yaml:59]: Good catch, I'd also say it'd be good to have the others as (even commented) env variables, so users can override them in individual ISVCs without having to touch the entrypoint line there

**Comment by Xaenalt** [comment]
Added a few more notes to my comments (message here for timestamp tracking)

File: docs/caikit-servingruntime.yaml
Comment (Xaenalt) [reply docs/caikit-servingruntime.yaml:59]: Yeah, overall, let's revert the cmdline args change, and just leave them as env vars commented so we can revisit in a subsequent PR, once that's done I'll merge this

File: Dockerfile
Comment (dtrifiro) [reply Dockerfile:27]: We do need python in the deploy stage, and that is why I'm using `FROM base as deploy`, I doubt `git-lfs` and the grpc endpoints are actually required, but they were present initially

Ideally, the deploy stage could be really minimal:
```dockerfile
FROM registry.access.redhat.com/ubi8/ubi-minimal:latest as deploy
WORKDIR /caikit

RUN microdnf -y update && \
    microdnf -y install shadow-utils python39 && \
    microdnf clean all && \
    groupadd --system caikit --gid 1001 && \
    adduser --system --uid 1001 --gid 0 --groups caikit \
    --create-home --home-dir /caikit --shell /sbin/nologin \
    --comment "Caikit User" caikit

COPY --from=poetry-builder /tmp/poetry/.venv /caikit/

ENV VIRTUAL_ENV=/caikit
ENV PATH="$VIRTUAL_ENV/bin:$PATH"

USER caikit

ENV CONFIG_FILES=/caikit/config/caikit.yml
VOLUME ["/caikit/config/"]

CMD ["python",  "-m", "caikit.runtime.grpc_server"]
```

**Comment by dtrifiro** [comment]
Updated the images/tags names as discussed with @Xaenalt

**Review by Xaenalt — APPROVED** [review]

**Comment by dtrifiro** [comment]
> @Xaenalt , actually, it's this image that I'm using at the moment: `quay.io/opendatahub/caikit-tgis-serving:stable` will it be updated automatically following this merge? 🤔

yep

File: docs/caikit-isvc.yaml
Comment (bdattoma) [reply docs/caikit-isvc.yaml:18]: @Xaenalt so now you don't need 2 layer of directories anymore right?

what's `proto`?

File: docs/caikit-isvc.yaml
Comment (bdattoma) [inline docs/caikit-isvc.yaml:12]: do we need to update the KServe CRD as well? currently we ship v1beta1 @Xaenalt

Comment (Xaenalt) [reply docs/caikit-isvc.yaml:12]: I think we want to align this with the current KServe version, so I think we'd want this to be v1beta1

**Comment by bdattoma** [comment]
@dtrifiro so, this PR is implementing only 1 scenario out of 3 (i.e., caikit and tgis same pod but different container), right? If that so, would you mind change the title and description to reflect that?

**Pull request merged**
